write_hooks_test.go

```diff
@@ -0,0 +1,88 @@
+package qfs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+	"testing"
+)
+
+func TestWriteHooks(t *testing.T) {
+	ctx := context.Background()
+	fs := NewMemFS()
+	bHash := ""
+
+	rewriteB := func(ctx context.Context, f File, pathMap map[string]string) (io.Reader, error) {
+		hContents, err := fs.Get(ctx, pathMap["/a/d.txt"])
+		if err != nil {
+			return nil, err
+		}
+		hData, err := ioutil.ReadAll(hContents)
+		if err != nil {
+			return nil, err
+		}
+		return strings.NewReader("APPLES" + string(hData)), nil
+	}
+
+	getBHash := func(ctx context.Context, f File, pathMap map[string]string) (io.Reader, error) {
+		bHash = pathMap["/a/b.txt"]
+		return f, nil
+	}
+
+	root := NewMemdir("/a",
```

Comment:
Please add a comment explaining the semantics here. If I am understanding correctly, "b.txt" depends on "d.txt", and rewrites it's contents by replacing "foo" with the results of function `rewriteB`.

It looks like "c.txt" isn't used in the test, and exists just to run a hook when "b.txt" finishes writing, in order to assign `bHash`. Maybe a useful future feature would be a standalone hook that runs when an AddFile operation finishes, but doesn't need it's own file (like "c.txt" here).